File: MatlabCode/MutualInfo.m
function [HA,HB,HAB,IAB]=MutualInfo(SigmaA,SigmaB,SampleNum,K)
% Input: 
% SigmaA: the covariance matrix Sigma of network A
% SigmaB: the covariance matrix Sigma of network B
% SampleNum: the number of samples in random sample generation
% K: the K value of KNN-based entropy estimation

% Output:
% HA: the entropy of network A
% HB: the entropy of network B
% HAB: the joint entropy of (A,B)
% IAB: the mutual information between network A and network B

[SampleA] = mvg(zeros(size(SigmaA,1),1),SigmaA,SampleNum); % Generate k=SampleNum random samples 
[SampleB] = mvg(zeros(size(SigmaB,1),1),SigmaB,SampleNum); % Generate k=SampleNum random samples
JointSamples=[SampleA',SampleB'];

HA=size(SigmaA,1)/2+size(SigmaA,1)/2*log(2*pi)+1/2*sum(log(eig(SigmaA)));
HB=size(SigmaB,1)/2+size(SigmaB,1)/2*log(2*pi)+1/2*sum(log(eig(SigmaB)));
HAB=EntropyEstimation(JointSamples,K);

IAB=(HA+HB-HAB)*(HA+HB>=HAB);

File: PythonCode/toolkit/mutual_info.py
import numpy as np

from .utils import multivar_gaussian_rand_num_generator, entropy_estimation


def mutual_infomation(sigma_a, sigma_b, sample_num, k):
    """
    Input:
    sigma_a: the covariance matrix Sigma of network a
    sigma_b: the covariance matrix Sigma of network b
    sample_num: the number of samples in random sample generation
    k: the number of nearest neighbors in KNN-based entropy estimation

    Output:
    mi: the mutual information between a and b
    """

    sample_a = multivar_gaussian_rand_num_generator(np.zeros(sigma_a.shape[0]), sigma_a, sample_num)[0]
    sample_b = multivar_gaussian_rand_num_generator(np.zeros(sigma_b.shape[0]), sigma_b, sample_num)[0]
    joint_samples = np.concatenate((sample_a, sample_b), axis=0).T
    
    h_a = (1.0 + np.log(2 * np.pi)) * sigma_a.shape[0] / 2.0 + np.sum(np.log(np.linalg.eigvals(sigma_a))) / 2.0
    h_b = (1.0 + np.log(2 * np.pi)) * sigma_b.shape[0] / 2.0 + np.sum(np.log(np.linalg.eigvals(sigma_b))) / 2.0
    h_ab = entropy_estimation(joint_samples, k)

    h_a = np.real((1.0 + np.log(2 * np.pi)) * sigma_a.shape[0] / 2.0 + np.sum(np.log(np.linalg.eigvals(sigma_a))) / 2.0)
    h_b = np.real((1.0 + np.log(2 * np.pi)) * sigma_b.shape[0] / 2.0 + np.sum(np.log(np.linalg.eigvals(sigma_b))) / 2.0)

    if h_a<0.0:
        h_a=0.0

    if h_b<0.0:
        h_b=0.0

    h_ab = entropy_estimation(joint_samples, k)

    mi = np.min([np.max([h_a + h_b - h_ab, 0]),h_a,h_b])

    return h_a, h_b, h_ab, mi
